byway/src/main/java/com/google/sps/servlets/StopsServlet.java

```diff
@@ -26,26 +37,39 @@
 /** Servlet that handles data for stops */
 @WebServlet("/api/stop")
 public final class StopsServlet extends HttpServlet {
-  @GuardedBy("this")
-  private ArrayList<String> stops = new ArrayList<String>();
+  private static final Type ARRAYLIST_STRING = new TypeToken<ArrayList<String>>() {}.getType();
+  private static final FluentLogger logger = FluentLogger.forEnclosingClass();
 
   private final Gson gson = new Gson();
-  // Implement datastore in routepage branch
+  private final DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+  private final Key key = KeyFactory.createKey(Trip.KIND, 1234);
 
+  /* Passes saved destinations stops (if any) to be shown in the schedule panel */
   @Override
-  public synchronized void doGet(HttpServletRequest request, HttpServletResponse response)
-      throws IOException {
-    Gson gson = new Gson();
-    String json = gson.toJson(stops);
-
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    List<String> stops = Collections.emptyList();
+    try {
+      Entity entity = datastore.get(key);
+      stops = (ArrayList<String>) entity.getProperty("destinations");
+    } catch (EntityNotFoundException e) {
+      logger.atInfo().withCause(e).log(
+          "Could not retrieve Entity for Trip with key %s while trying to get the stops", key);
+    }
     response.setContentType("application/json;");
-    response.getWriter().println(json);
+    response.getWriter().println(gson.toJson(stops));
   }
 
+  /* Modifies the destinations array of Trip entity in datastore */
   @Override
-  public synchronized void doPost(HttpServletRequest request, HttpServletResponse response)
-      throws IOException {
-    String stop = request.getParameter("text");
-    stops.add(stop);
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    ArrayList<String> stops = gson.fromJson(request.getReader(), ARRAYLIST_STRING);
+    try {
+      Entity entity = datastore.get(key);
+      entity.setProperty("destinations", stops);
+      datastore.put(entity);
+    } catch (EntityNotFoundException e) {
+      logger.atInfo().withCause(e).log(
+          "Could not retrieve Entity for Trip with key %s while trying to update the stops", key);
+    }
```

Comment:
It would be good to set status codes on success and failure, so you give the client code some idea what the error is, or what the result is. With `response.setStatus`, you can use the status code constants from `HttpServletResponse`, for example when the entity is not found, you could send `SC_NOT_FOUND`, and when a request succeeds but you have nothing to actually send back, you can send `SC_NO_CONTENT`.

Reply:
Okay, I've added this setStatus() to both methods.